net/l2: openthread: Add support for automatic joiner start

takumiando opened the pull request:
We sometimes want to join a device to OpenThread mesh automatically.
This commit adds supports to start joiner automatically by Kconfig.

The default value of CONFIG_OPENTHREAD_JOINER_PSKD is based on this page:
https://codelabs.developers.google.com/codelabs/openthread-hardware/

Signed-off-by: Takumi Ando <takumi.ando@atmark-techno.com>

takumiando commented:
If this config is enabled, Zephyr starts joiner and join after power on like this:

```
***** Booting Zephyr OS zephyr-v1.13.0-4179-g1307ea3ddb *****
[00:00:00.000,000] <inf> ieee802154_nrf5: nRF5 802154 radio initialized
[00:00:00.000,000] <wrn> net_dns_resolve: Cannot initialize DNS resolver (-22)
[00:00:00.174,255] <inf> net_l2_openthread: OpenThread version: OPENTHREAD/20180926-00241-g30188869; NONE; Feb 28 2019 04:27:07
[00:00:00.174,285] <inf> net_l2_openthread: Network name:   degu
[00:00:00.174,957] <inf> net_l2_openthread: State changed! Flags: 0x040f1219 Current role: 0
[00:00:00.175,109] <inf> net_l2_openthread: State changed! Flags: 0x00008000 Current role: 0
[00:00:00.175,262] <inf> net_l2_openthread: State changed! Flags: 0x00000002 Current role: 0
[00:00:00.175,354] <inf> net_l2_openthread: State changed! Flags: 0x00000001 Current role: 0
[00:00:00.175,445] <inf> net_l2_openthread: State changed! Flags: 0x00000008 Current role: 0
[00:00:00.175,811] <inf> net_l2_openthread: Start joiner success
[00:00:00.181,732] <inf> net_config: Initializing network
[00:00:00.181,884] <inf> net_l2_openthread: State changed! Flags: 0x00000001 Current role: 0
[00:00:07.252,380] <inf> net_l2_openthread: State changed! Flags: 0x00010000 Current role: 0
[00:00:13.781,555] <inf> net_l2_openthread: State changed! Flags: 0x00181111 Current role: 0
[00:00:13.882,232] <inf> net_l2_openthread: Join success
[00:00:13.883,026] <inf> net_l2_openthread: State changed! Flags: 0x0000801f Current role: 1
[00:00:15.015,655] <inf> net_l2_openthread: State changed! Flags: 0x006012a4 Current role: 2
[00:01:14.903,686] <inf> net_l2_openthread: State changed! Flags: 0x00000064 Current role: 3
```

Of course, we must add the device(new joiner) to mesh by commissioner before power on.

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -334,9 +355,31 @@ static int openthread_init(struct net_if *iface)
+	otError error;
+
+	error = otJoinerStart(ot_context->instance, CONFIG_OPENTHREAD_JOINER_PSKD, NULL, PACKAGE_NAME,
+			CONFIG_OPENTHREAD_PLATFORM_INFO, PACKAGE_VERSION, NULL, &ot_joiner_start_handler, ot_context);
```
This looks like a line that is over 80 chars long, I wonder why checkpatch has not complained about it.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -334,9 +355,31 @@ static int openthread_init(struct net_if *iface)
 {
 	openthread_init(iface);
+
+#if defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)
```
I'm afraid the logic here is too simple. With this approach, application will try to initiate the commissioning on each reboot.

There's an API function to verify that , check https://github.com/openthread/openthread/blob/bd41c63a0a041a4a43a5deab421d719a7011d5ca/include/openthread/dataset.h#L275

In case there's a valid network stored, you'd only need to start it with `otThreadSetEnabled`, otherwise initiate the joining process.

And I'd move this piece within `openthread_init` to keep the network starting logic in a single place.

takumiando replied:
Thanks for your reply!

Which do you mean:

```
otThreadSetEnabled(..., true);
if (!otDatasetIsCommissioned()) {
  otThreadSetEnabled(..., false);
  otJoinerStart(...);
}

callback() {
  otThreadSetEnabled(..., true);
}
```

or,

```
if (!otDatasetIsCommissioned()) {
  otJoinerStart(...);
}

callback() {
  otThreadSetEnabled(..., true);
}
```
?
Should I enable thread absolutely before call otDatasetIsCommissioned() to get network validation?

takumiando replied:
I don't understand why otDatasetIsCommissioned() returns false absolutely even zephyr joined successful and rebooted or not in my environment...

rlubos replied:
I meant something like this:
```
if (!otDatasetIsCommissioned()) {
    otJoinerStart(...);
} else {
    otThreadSetEnabled(..., true);
}

callback() {
    otThreadSetEnabled(..., true);
}
```

`otThreadSetEnabled(..., true);` simply starts the Thread network. It will use the parameters that we obtained during commissionning process (that's why we call it in the joiner callback). But the same parameters are (or at least should, if not we have a bug somewhere) stored in the flash, hence after reboot there's no need to go through the commissioning process again, we can start rightaway.

> And I don't understand why otDatasetIsCommissioned() returns false absolutely even zephyr joined successful and rebooted or not in my environment...

Hmm that's weird, let my try your PR on desk setup, perhaps there's something fishy going on. What hardware do you use?

takumiando replied:
I applied changes you told and added a patch for test:
https://github.com/takumiando/zephyr/commits/openthread_autostart_joiner_test

I'm using nrf52840_pca10056 board, and use sample/net/socket/echo_client with test patch.

```
$ cd sample/net/socket/echo_client
$ mkdir build && cd build
$ cmake -GNinja -DBOARD=nrf52840_pca10056 -DOVERLAY_CONFIG=overlay-ot.conf ..
$ ninja
```

And run it, my board joins to my network:

```
***** Booting Zephyr OS v1.14.0-rc1-625-g3a012c9502e4 *****
[00:00:14.357,666] <inf> net_l2_openthread: State changed! Flags: 0x00000010 Current role: 0
[00:00:14.451,110] <inf> net_l2_openthread: Join success
[00:00:14.454,010] <inf> net_l2_openthread: State changed! Flags: 0x0000903f Current role: 1
[00:00:16.446,075] <inf> net_l2_openthread: State changed! Flags: 0x00000040 Current role: 1
[00:00:17.412,109] <inf> net_l2_openthread: State changed! Flags: 0x000012a4 Current role: 2
```

Then, I reset my board but it tries to start joiner again...

```
***** Booting Zephyr OS v1.14.0-rc1-625-g3a012c9502e4 *****
[00:00:17.249,816] <inf> net_l2_openthread: State changed! Flags: 0x00000010 Current role: 0
[00:00:17.345,092] <inf> net_l2_openthread: Join success
[00:00:17.348,052] <inf> net_l2_openthread: State changed! Flags: 0x0000903f Current role: 1
[00:00:19.340,057] <inf> net_l2_openthread: State changed! Flags: 0x00000040 Current role: 1
[00:00:20.800,231] <inf> net_l2_openthread: State changed! Flags: 0x000012a4 Current role: 2
```

My understanding is otDatasetIsCommissioned() returns false both these cases.
I can't understand why it does... 

I dumped and compared flash on my board by nrfjprog at before joined and after.

```
--- before_join	2019-03-01 11:12:26.249137691 +0900
+++ after_join	2019-03-01 11:13:10.633884614 +0900
@@ -64766,20 +64766,20 @@
 0x000FCFD0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FCFE0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FCFF0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD000: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD010: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD020: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD030: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD040: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD050: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD060: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD070: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD080: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD090: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0A0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0B0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0C0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0D0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
+0x000FD000: C5EE BE5C 0003 00F4 0024 0000 0000 0000   |..\.....$.......|
+0x000FD010: 0001 0000 03E8 0000 03E8 0000 0000 0000   |................|
+0x000FD020: 0000 0000 0000 0000 0000 0000 0000 0000   |................|
+0x000FD030: 0001 00F4 005E CE82 0300 0000 3514 0006   |....^........5..|
+0x000FD040: 0004 FF1F 02E0 0508 5542 A7AF 12E8 0767   |........BU....g.|
+0x000FD050: FD08 4205 AF55 00A7 0500 DE10 2B7B 789D   |...BU.......{+.x|
+0x000FD060: D48D 1A57 F3E2 A543 0E4D 03F6 6404 6765   |..W...C.M....deg|
+0x000FD070: 0175 AB02 04CD 0010 0000 0000 0000 0000   |u...............|
+0x000FD080: 0000 0000 0000 0C00 0203 FFA0 080E 0000   |................|
+0x000FD090: 0000 0000 0000 FFFF 0004 00F4 0008 724C   |..............Lr|
+0x000FD0A0: F456 724C 1925 A668 0003 00F4 0024 724C   |V.Lr%.h.....$.Lr|
+0x000FD0B0: 0F02 9C1E 0001 0000 03EA 0000 03E8 0000   |................|
+0x000FD0C0: 9A22 5306 5866 70AA 8A27 92E9 05B2 97AD   |"..SfX.p'.......|
+0x000FD0D0: F91C 6A10 FFFF FFFF FFFF FFFF FFFF FFFF   |...j............|
 0x000FD0E0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FD0F0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FD100: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
```

otDatasetIsCommissioned() checks this difference region, is that right?

rlubos replied:
Hi @takumiando , thanks for investigating! `otDatasetIsCommissioned` checks the Active Dataset loaded into RAM from flash memory during OpenThread instance initialization. I did some debugging on my own, and realized that `otDatasetIsCommissioned` returns true just after `otInstanceInitSingle`. Yet, when called just before joiner start, it reutrns false. Therefore it seems that either something corrupts the dataset in RAM, or it is discarded for whatever reason and overwritten. I"m going to investigate the issue further and hopefully propose a patch to fix it.

rlubos replied:
Ok, found the showstopper.

> or it is discarded for whatever reason and overwritten

Apparently that was the case. I missed the fact, that network name is also obtained during commissioning, so we should not set it manually with `otThreadSetNetworkName`. Otherwise, the dataset is invalidated:
https://github.com/openthread/openthread/blob/8c13df5a51f4b0428d73f0bc57a01506c178e421/include/openthread/thread.h#L459

Therefore, bundling `otThreadSetNetworkName` with other parameters underneath 
`#if !defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)` solves the issue! Just note, that OT flash data region is not cleared when you program the application - you may want to erase the board with `nrfjprog -e` to test the joining procedure.

takumiando replied:
Thanks for investigating!
I confirmed that zephyr does not start joiner after reboot.

I fixed this issue and force-pushed to the same branch!

rlubos reviewed (changes requested):
Looks like a step in right direction 👍

rlubos commented on subsys/net/l2/openthread/Kconfig:
```diff
@@ -150,6 +150,24 @@ config OPENTHREAD_JOINER
+	  Pre Shared Key for the Device to start joiner
+
+config OPENTHREAD_PLATFORM_INFO
+	string "Platform infomation of OpenThread"
```
Just noticed, please make this `Platform information for OpenThread` here and below

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -302,15 +338,22 @@ static int openthread_init(struct net_if *iface)
 	otIp6SetEnabled(ot_context->instance, true);
+
+#if !defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)
 	otThreadSetEnabled(ot_context->instance, true);
```
Please move this below, it'll be easier to follow how network starts if we keep the logic in on place. I mean something like this:
```
#if !defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)
    otThreadSetEnabled(ot_context->instance, true);
#else
    openthread_joiner_autostart(iface);
#endif
```

takumiando replied:
Thank you! I fixed it.

rlubos reviewed (changes requested):
Looks good, thanks. One last request and I think we are good to go.

takumiando committed:
net/l2: openthread: Add support for automatic joiner start

We sometimes want to join a device to OpenThread mesh automatically.
This commit adds supports to do by Kconfig.

The default of CONFIG_OPENTHREAD_JOINER_PSKD is based on this page:
https://codelabs.developers.google.com/codelabs/openthread-hardware/

Signed-off-by: Takumi Ando <takumi.ando@atmark-techno.com>

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -279,6 +297,24 @@ enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
+	otError error;
+
+	if (!otDatasetIsCommissioned(ot_context->instance)) {
+		error = otJoinerStart(ot_context->instance, CONFIG_OPENTHREAD_JOINER_PSKD, NULL,
```
Can you reformat these three lines, they seem to be over 80 char limit?

takumiando replied:
Thanks! I fixed it.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -279,6 +297,28 @@ enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
 	return NET_CONTINUE;
 }
 
+void openthread_joiner_autostart(struct net_if *iface)
```
One more issue spotted, this does not seem to build in default configuration, when joiner autostart is disabled:
```
/home/rolu/repos/zephyrproject/zephyr/subsys/net/l2/openthread/openthread.c:306:47: error: ‘CONFIG_OPENTHREAD_JOINER_PSKD’ undeclared (first use in this function)
   error = otJoinerStart(ot_context->instance, CONFIG_OPENTHREAD_JOINER_PSKD, NULL,
```

I guess we should conditionally disable building of this function with `#if defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)`

takumiando replied:
Oh sorry, I've forgot it...
I fixed it.

rlubos reviewed (approved):
Thanks, LGTM 👍 Just make sure to fix checkpatch errors, otherwise CI will not let this PR in.

takumiando commented:
> Thanks, LGTM  Just make sure to fix checkpatch errors, otherwise CI will not let this PR in.

Fixed it!

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -279,6 +297,29 @@ enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
+
+	if (!otDatasetIsCommissioned(ot_context->instance)) {
+		error = otJoinerStart(ot_context->instance,
+					CONFIG_OPENTHREAD_JOINER_PSKD, NULL,
```
Sorry for nitpicking but I just noticed that there is some style issue still present.
Please align the parameters to same column.

takumiando replied:
Thanks! I fixed it.

jukkar commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -279,6 +297,29 @@ enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
+					NULL, &ot_joiner_start_handler,
+					ot_context);
+
+		if (error != OT_ERROR_NONE)
```
And we always put {} even for one liners like this.

takumiando replied:
Thanks! I fixed it.

mike-scott reviewed (approved):
Nice addition
